contrib/dump-mtd-ifd.py

```diff
@@ -0,0 +1,177 @@
+        self.size: int = size
+
+    def __str__(self) -> str:
+        return (
+            "IfdPartition("
+            f"region={IfdPartition.REGION_NAMES[self.region]}, "
+            f"offset=0x{self.offset:x}, "
+            f"size=0x{self.size:x})"
+        )
+
+
+def _read_partitions(f: io.BufferedReader) -> bytearray:
+    f.seek(0)
+    blob_ifd: bytes = f.read(0x1000)
+    (
+        signature,
+        descriptor_map0,
+        descriptor_map1,
+        descriptor_map2,
+    ) = struct.unpack_from("<16xIIII", blob_ifd, offset=0)
+    if signature != 0x0FF0A55A:
+        print(f"Not IFD signature 0x0FF0A55A, got 0x{signature:X}")
+        sys.exit(1)
+
+    # read out the descriptor maps
+    print(f"descriptor_map0=0x{descriptor_map0:X}")
+    print(f"descriptor_map1=0x{descriptor_map1:X}")
+    print(f"descriptor_map2=0x{descriptor_map2:X}")
+
+    # default to all partitions
+    num_regions = (descriptor_map0 >> 24) & 0b111
+    if num_regions == 0:
+        num_regions = 10
+    print(f"num_regions=0x{num_regions:X}")
+
+    # read out FREGs
+    flash_region_base_addr = (descriptor_map0 >> 12) & 0x00000FF0
+    print(f"flash_region_base_addr=0x{flash_region_base_addr:X}")
+    flash_descriptor_regs: List[int] = []
+    for region in range(num_regions):
+        flash_descriptor_regs.append(
+            struct.unpack_from(
+                "<I", blob_ifd, offset=flash_region_base_addr + (region * 4)
+            )[0]
+        )
+    for region in range(num_regions):
+        print(f"flash_descriptor_reg{region}=0x{flash_descriptor_regs[region]:X}")
+
+    # parse each partition
+    fregs: List[IfdPartition] = []
+    for i in range(num_regions):
+        freg_base: int = (flash_descriptor_regs[i] << 12) & 0x07FFF000
+        freg_limit: int = ((flash_descriptor_regs[i] >> 4) & 0x07FFF000) | 0x00000FFF
+
+        # invalid
+        if freg_base > freg_limit:
+            continue
+
+        fregs.append(
+            IfdPartition(region=i, offset=freg_base, size=freg_limit - freg_base)
+        )
+
+    # create a binary blob big enough
+    image_size: int = 0
+    for freg in fregs:
+        if freg.offset + freg.size > image_size:
+            image_size = freg.offset + freg.size
+    print(f"image_size=0x{image_size:x}")
+    blob: bytearray = bytearray(image_size)
+
+    # copy each partition
+    for freg in fregs:
+        print("reading...", freg)
+        try:
+            f.seek(freg.offset)
+            blob[freg.offset : freg.offset + freg.size] = f.read(freg.size)
+        except OSError as e:
+            print(f"failed to read: {e}")
+    return blob
+
+
+def _read_device_to_file(devname: str, filename: Optional[str]) -> None:
+    # grab system info from sysfs
+    if not filename:
+        filename = ""
+        for sysfs_fn in [
+            "/sys/class/dmi/id/sys_vendor",
+            "/sys/class/dmi/id/product_family",
+            "/sys/class/dmi/id/product_sku",
+        ]:
```

Comment:
I recall Dell does it differently than Lenovo; you might want to get `product_name` too

Reply:
Good catch, thanks.